wrappers/azurekeyvault/azurekeyvault.go

```diff
@@ -261,26 +272,295 @@ func (v *Wrapper) Decrypt(ctx context.Context, in *wrapping.EncryptedBlobInfo, a
+	kekBundle, err := v.client.CreateKey(ctx, v.buildBaseURL(), kekName, keyvault.KeyCreateParameters{
+		Kty:     keyvault.RSAHSM,
+		KeySize: to.Int32Ptr(2048),
+		KeyOps:  &[]keyvault.JSONWebKeyOperation{"import"},
+		KeyAttributes: &keyvault.KeyAttributes{
+			Expires: &exp,
+		},
+	})
+	if err != nil || kekBundle.Key == nil {
+		return "", fmt.Errorf("error generating KEK: %w", err)
+	}
+	defer func() {
+		v.client.DeleteKey(ctx, v.buildBaseURL(), kekName)
+	}()
+
+	// Parse the RSA public key of the KEK
+	kekPubKey, err := jwkToRSAPublicKey(kekBundle.Key)
+	if err != nil {
+		return "", err
+	}
+
+	// Produce the target key plaintext
+	var targetKey []byte
+	switch key.Type {
+	case wrapping.RSA2048, wrapping.RSA3072, wrapping.RSA4096:
+		// For an RSA key, the private key ASN.1 DER encoding [RFC3447] wrapped in PKCS#8 [RFC5208]
+		if targetKey, err = x509.MarshalPKCS8PrivateKey(key.Material.RSAKey); err != nil {
+			return "", err
+		}
+	}
+
+	// Produce the wrapped key material needed for import
+	wrappedKeyMaterial, err := v.wrapTargetKey(kekPubKey, targetKey)
+	if err != nil {
+		return "", err
+	}
+
+	// Create the key transfer blob
+	b := keyTransferBlob{
+		SchemaVersion: "1.0.0",
+		Header: keyTransferBlobHeader{
+			Alg: "dir",
+			Enc: "CKM_RSA_AES_KEY_WRAP",
+			Kid: to.String(kekBundle.Key.Kid),
+		},
+		CipherText: wrappedKeyMaterial,
+		Generator:  "Hashicorp",
+	}
+	ktb, err := json.Marshal(b)
+	if err != nil {
+		return "", err
+	}
+	ktbEncoded := base64.RawURLEncoding.EncodeToString(ktb)
+
+	// Upload the key transfer blob to import the key
+	kty := v.keyTypeToKty(key.Type)
+	ops := v.keyPurposesToKeyOps(key.Purposes)
+	imported, err := v.client.ImportKey(ctx, v.buildBaseURL(), name, keyvault.KeyImportParameters{
+		Key: &keyvault.JSONWebKey{
+			Kty:    kty,
+			T:      to.StringPtr(ktbEncoded),
+			KeyOps: to.StringSlicePtr(ops),
+		},
+	})
+	if err != nil {
+		return "", fmt.Errorf("error importing key: %w", err)
+	}
+	if imported.Key == nil || imported.Key.Kid == nil {
+		return "", errors.New("imported key is missing identifier")
+	}
+
+	// Return the version ID generated for the imported key
+	return parseKeyVersion(to.String(imported.Key.Kid)), nil
 }
 
-func (v *Wrapper) RotateKey(_ context.Context, _ string, _ wrapping.KMSKey) (string, error) {
-	return "", nil
+func (v *Wrapper) RotateKey(ctx context.Context, name string, key wrapping.KMSKey) (string, error) {
+	// Check that the key exists before importing a new version
+	if _, err := v.client.GetKey(ctx, v.buildBaseURL(), name, ""); err != nil {
+		return "", err
+	}
+
+	// Rotation works by importing key material into an existing key
+	return v.ImportKey(ctx, name, key)
 }
 
-func (v *Wrapper) DeleteKey(_ context.Context, _ string) (bool, error) {
-	return false, nil
+func (v *Wrapper) DeleteKey(ctx context.Context, name string) (bool, error) {
+	res, err := v.client.DeleteKey(ctx, v.buildBaseURL(), name)
+	return res.StatusCode != http.StatusNotFound, err
```

Comment:
Can `v.client.DeleteKey` return both a non-empty `res` and `err` at the same time? Asking because both values are being checked at the same time in tests.

Reply:
Yep, it can return both non-empty since `res` isn't a pointer. After looking at this closer, I saw that the embedded struct holding the `StatusCode` is a pointer, so I added a nil check and some comments to explain this in 9037b6e. I also updated the test to check what's returned independently.